fix(messaging, android): prevent OOM by limiting stored notifications

sean5940 opened the pull request:
## Description

This PR addresses an `OutOfMemoryError` (OOM) crash on Android caused by `ReactNativeFirebaseMessagingStoreImpl` storing up to 100 notifications in `SharedPreferences`.

When a device receives many notifications with large payloads, the `SharedPreferences` file grows significantly. Since `SharedPreferences` loads the entire file into memory, this can exceed the heap size on some devices, leading to a crash.

**Related Issue:**
Closes #8770

## Changes

1.  **Reduced Default Limit**: The default `MAX_SIZE_NOTIFICATIONS` has been reduced from **100** to **20**.
    - *Reasoning*: `getInitialNotification` typically only needs the single notification that launched the app. A buffer of 20 is sufficient for most use cases while significantly reducing memory pressure.

2.  **Configurable Limit**: Added support for configuring the limit via `AndroidManifest.xml`.
    - Developers can now set `rn_firebase_messaging_max_stored_notifications` in their `AndroidManifest.xml` to adjust this limit without forking the library.
    ```xml
    <meta-data
      android:name="rn_firebase_messaging_max_stored_notifications"
      android:value="50" />
    ```

3.  **Safety Cap**: Added a hard cap of **100** to the configurable limit.
    - Even if a developer sets the value higher in the manifest, it will be capped at 100 to prevent re-introducing the OOM risk.

4.  **Aggressive Cleanup**: Changed the cleanup logic from an `if` statement to a `while` loop.
    - *Reasoning*: If a user already has >20 notifications stored (e.g., 100 from the previous version), the previous logic only removed *one* old notification per new message. The `while` loop ensures the storage is immediately drained down to the limit upon receiving the next message.

## Verification

- Verified locally that `getInitialNotification` still functions correctly with the reduced limit.
- Verified that the limit is correctly read from `AndroidManifest.xml`.
- Verified that the safety cap prevents values > 100.
- Verified that the `while` loop correctly cleans up excess notifications.

## Stack Trace (Reference)

```java
java.lang.OutOfMemoryError: Failed to allocate a 24 byte allocation with 425280 free bytes and 415KB until OOM...
    at android.app.SharedPreferencesImpl$EditorImpl.commitToMemory(SharedPreferencesImpl.java:539)
    at io.invertase.firebase.messaging.ReactNativeFirebaseMessagingStoreImpl.storeFirebaseMessage(ReactNativeFirebaseMessagingStoreImpl.java:43)
```

mikehardy commented on packages/messaging/android/src/main/java/io/invertase/firebase/messaging/ReactNativeFirebaseMessagingStoreImpl.java:
```diff
@@ -18,8 +18,31 @@
   private static final String S_KEY_ALL_NOTIFICATION_IDS = "all_notification_ids";
-  private static final int MAX_SIZE_NOTIFICATIONS = 100;
   private final String DELIMITER = ",";
+  private static int MAX_SIZE_NOTIFICATIONS = 20;
```
I don't believe changing this is a good idea - not for 100% technical reasons - 100 may not be practical for every case - you obviously have an OOM so this is critical to address somehow - but this feature has existed for 5 years (ref: https://github.com/sean5940/react-native-firebase/commit/a7cafc92f4250d2ccb16cb181a4029f14b76b8ca) and you are the first report of a problem.

So while 100 may be impractical sometimes, it's once in 5 years out of maybe tens of thousands of apps and millions and millions of users.

Weigh that versus even the slim possibility of some app relying on this behavior (or some unknown amount less than 100 and greater than 20) and that implies this would be a breaking change.

I don't think a fix here is worth a breaking change given it has existed for 5 years without any other issues.

So I am a "no" on changing this constant

mikehardy commented on packages/messaging/android/src/main/java/io/invertase/firebase/messaging/ReactNativeFirebaseMessagingStoreImpl.java:
```diff
@@ -18,8 +18,31 @@
+  private int getMaxNotificationSize() {
+    if (isInitialized) return MAX_SIZE_NOTIFICATIONS;
+    try {
+      android.content.Context context =
```
I don't like the fully qualified package names here vs imports alphabetic at the top and just the names here. Fully qualified package names when they are not critical to disambiguate similarly named objects do not match style

mikehardy reviewed (changes requested):
I mentioned in more detail in specific comments but a summary:

1- I really like the idea of making the serialized notification count configurable

But, I believe strongly this should use our existing code though, and if you want to make existing boot-crash-loop devices recoverable you'll need to add error handling to that common code

2- I do not believe we should change the default, we should avoid a potentially breaking change

3- I like the trim-until-we-are-the-value vs one-at-a-time change

I think there's the core of something that can go in here if you work through the info I put out for item 1 here in this list

mikehardy commented:
@sean5940 your analysis appears correct, regardless of the other feedback I gave, your conversion of the "if" to a "while", and adding an `apply()` there is an immediate bug fix. I will post a PR and release it independently of this here so that it not blocked and the fix can be out with no delay

mikehardy commented:
@sean5940 I can merge and release the PR quickly, but review always helps. If you could take a look?

- #8776

I would still like to pursue the idea of making the storage limit configurable here, and working through the idea of OOM recovery by catching an OOM error. Although I think the 5th commit in my PR - "purge old messages before storing new" may actually recover people sufficiently,.

sean5940 commented:
> This was mostly superceded by #8776 which I wanted to get out as soon as possible because it was in "crash fix" territory
> 
> The remaining part of this is the configurable maximum message parameter - what do you think @sean5940 is it still worth it to implement that, or do you want to see if #8776 clears out the problem?

@mikehardy 

One additional concern I have is that once the file size gets very large, the chances of running into an OOM error go up significantly. I’ve been thinking about how we could handle this scenario. Do you have any suggestions or approaches that might help mitigate it?

sean5940 committed:
fix(messaging): honor custom max stored notifications

sean5940 committed:
fix(messaging): enforce min stored notification limit

sean5940 committed:
feat(messaging): add stored notification config

sean5940 commented on packages/messaging/android/src/main/java/io/invertase/firebase/messaging/ReactNativeFirebaseMessagingStoreImpl.java:
```diff
@@ -18,8 +18,31 @@
+      if (appInfo.metaData != null) {
+        int configValue =
+            appInfo.metaData.getInt("rn_firebase_messaging_max_stored_notifications", 20);
+        MAX_SIZE_NOTIFICATIONS = Math.min(configValue, 100);
```
Thanks for the detailed feedback. I’ve updated the implementation to read the configuration value using the approach you mentioned at the beginning (SharedPreferences → firebase.json → AndroidManifest).

[41ca8ef](https://github.com/invertase/react-native-firebase/pull/8771/commits/41ca8efb3e20177a93dae834fa808209d7a3603b)

mikehardy reviewed (changes requested):
Thanks for the config change - nice to give people control over this one
As mentioned I don't think a clamp is a positive move but that's an easy change

And if you wanted to implement an all-messageIds-iterator in a separate method like `preenMessageStorage` or similar then call it before storing a new message, I'd be open to that

sean5940 committed:
fix(messaging, android): update config key name and remove clamping per review

- Rename config key from 'rn_firebase_messaging_max_stored_notifications' to 'messaging_max_stored_notifications' for consistency
- Remove clamping logic (min/max restrictions) to allow developers full control
- Add logging for final setting value and source (SharedPreferences/firebase.json/AndroidManifest/default) for diagnostics
- Update firebase-schema.json: remove minimum/maximum constraints and update description
- Update tests/firebase.json to use new key name

Addresses review feedback: developers should have full control over the limit without artificial restrictions, with appropriate warnings in documentation.

sean5940 committed:
fix(app, android): handle Integer values in ReactNativeFirebaseMeta.getAll()

The getAll() method was only handling null, String, and Boolean values when converting metadata to a WritableMap. With the new getIntValue() method supporting integer configuration values (e.g., messaging_max_stored_notifications), Integer values in metadata were being silently skipped.

Added Integer type check to properly include Integer values in the returned WritableMap.

mikehardy reviewed (commented):
most recent two commits look good on a scan - thanks for the bugfix on Meta integer key handling - approving workflows now as the code looks good and just needs a review of all commits together as a batcnh
